src/main/java/nextstep/qna/service/QnAService.java

```diff
@@ -24,26 +23,10 @@ public class QnAService {
     private DeleteHistoryService deleteHistoryService;
 
     @Transactional
-    public void deleteQuestion(NsUser loginUser, long questionId) throws CannotDeleteException {
+    public void deleteQuestion(NsUser loginUser, long questionId) {
         Question question = questionRepository.findById(questionId).orElseThrow(NotFoundException::new);
-        if (!question.isOwner(loginUser)) {
-            throw new CannotDeleteException("질문을 삭제할 권한이 없습니다.");
-        }
-
-        List<Answer> answers = question.getAnswers();
-        for (Answer answer : answers) {
-            if (!answer.isOwner(loginUser)) {
-                throw new CannotDeleteException("다른 사람이 쓴 답변이 있어 삭제할 수 없습니다.");
-            }
-        }
-
-        List<DeleteHistory> deleteHistories = new ArrayList<>();
-        question.setDeleted(true);
-        deleteHistories.add(new DeleteHistory(ContentType.QUESTION, questionId, question.getWriter(), LocalDateTime.now()));
-        for (Answer answer : answers) {
-            answer.setDeleted(true);
-            deleteHistories.add(new DeleteHistory(ContentType.ANSWER, answer.getId(), answer.getWriter(), LocalDateTime.now()));
-        }
-        deleteHistoryService.saveAll(deleteHistories);
+        DeleteHistories deleteHistories = new DeleteHistories();
+        question.delete(loginUser, deleteHistories);
+        deleteHistoryService.saveAll(deleteHistories.getDeleteHistories());
```

Comment:
DeleteHistories 를 서비스가 생성할 필요가 있을까요? 🤔 

```suggestion
        DeleteHistories deleteHistories = question.delete(loginUser);
        deleteHistoryService.saveAll(deleteHistories.getDeleteHistories());
```

이렇게 객체에게 메세지를 보낸 후 메세지에 대한 결과를 받는 방식은 어떠신가요?

Reply:
> 서비스에 제가 생성 한 이유는 delete() 메서드를 호출한 반환타입이 DeleteHistories 인것이 부자연 스러워 그랬는데 return type과 메서드 이름에는 연관관계가 없어도 되는건가요?

부자연스러운 이유가 어떤것이었을까요??
매개변수로 DeleteHistories 를 받는것이 좀더 자연스러우셨을까요?
이유가 궁금합니다!

Reply:
사실 메서드 사용시 파라미터 개수가 적으면 적을수록 이해하기 쉽고 명확도는 올라가는것 같습니다.

그렇다 보니 민수님이 말하신 방식을 사용하는 것이 더 나은 선택 같네요 ㅎㅎ

감사합니다!

Reply:
저도 사실 어떻게 deleteHistories를 생성해서 넘길까 고민이 많았는데, 앞으로 파라미터로 넘기는 대신 반환타입을 사용하는 방식 쪽으로 작성해보갰습니다.

그것 보다 해당 deleteHistories 객체를 어디서 생성해야 할지가 애매하긴 한 것 같습니다.
상태값으로 가지는 것은 아닌것 같고, 메서드 파라미터로 넘겨 줘야 하는데, 위치를 선정하는데 있어서 애매한것 같아요.

Reply:
아니면 answer 쪽 가장 depth의 끝 부분에서 생성하여 반환타입으로 넘겨주는게 맞을까요?

Reply:
@juno-junho

```java
// Answer.java
public DeleteHistory delete() {
    return DeleteHistory.createAnswer(this.id, this.writer);
}

// Question.java
public DeleteHistories delete() {
    DeleteHistories deleteHistories = new DeleteHistories();
    for(Answer answer : answers) {
        deleteHistories.add(answer.delete());
    }
    deleteHistories.add(DeleteHistory.createQuestion(this.id, this.writer);
    return deleteHistories;
}

// DeleteHistory.java
public static DeleteHistory createQuestion(String id, NsUser writer){
    return new DeleteHistory(ContentType.Question, id, writer);
}

public static DeleteHistory createAnswer(String id, NsUser writer){
    return new DeleteHistory(ContentType.Answer, id, writer);
}
```

이렇게 할수 있을것 같아요! (IDE 아니고 깃허브에서 코딩하니 문법이 틀릴 수 있어요 😢  양해바랍니다)

그러면 서비스에선

```java
DeleteHistories deleteHistories = question.delete();
```

이렇게 되겠고, 코드를 한글로 읽는다 생각하면
질문을 삭제하면? 삭제 이력이 나온다. 로 바로 해석될 수 있지 않을까요??

Reply:
이런 설계는 가독성, 전체적 논리적 흐름을 고려해서 어느 클래스에 어떤 책임을 부여할지 결정하는 것인가요?